### PR TITLE
Origin/enh/character table

### PR DESCRIPTION
**Pull request by tweisser**
Very rudimentary implementation to get things started. I stopped because of the following issue:
>>>Commenting out line 142 returns the wrong characteristic polynomial because M[2] is wrong.

### CODE REVIEW

File: src/characters.jl [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    return z
+end
+
+m = Int(lcm(order.(G)))
+q = find_prime(m)
+
+
+@ncpolyvar T
+println(M[2])
+p = mod(characteristic_polynomial(M[2].m, T), q)
```
Comment (kalmarek) [inline src/characters.jl]: try `p = mod(characteristic_polynomial(M[2], T), q)`

`M[2]` is already an array, so use it as such, never use its internals: `M[2].m` is only initialized to correct values when accessed via `getindex`; `println` asks for values of `M[2]`, so then you get the correct answer;

we may change the name of field from `m` to `lazyclasscoeffs` or something like this.

why T is a `@ncpolyvar` (or what exactly it is?)

with `Nemo` I can do 
```julia
F = Nemo.GF(23)
Fx, x = F["x"]
NM = [matrix(F, m) for m in M]
for m in NM
    @show factor(minpoly(Fx, m))
end
```
```
factor(minpoly(Fx, m)) = 1 * (x+22)
factor(minpoly(Fx, m)) = 1 * (x) * (x+17) * (x+21) * (x+2) * (x+6)
factor(minpoly(Fx, m)) = 1 * (x) * (x+15) * (x+4)
factor(minpoly(Fx, m)) = 1 * (x+20) * (x+1)
factor(minpoly(Fx, m)) = 1 * (x) * (x+17) * (x+21) * (x+2) * (x+6)
```
unfortunately support for factorizations, etc in AA is poor as it's intended to be used as framework for `Nemo`.

**Comment by kalmarek** [comment]
Maybe we can sidestep polynomials and their factorisations, by looking at the eigenvalues instead. 
In AA there is already a `nullspace`:
```
  nullspace(M::AbstractAlgebra.MatElem{T}) where {T <: FieldElement}

  │  Return a tuple (\nu, N) consisting of the nullity \nu of M and a basis N
  │  (consisting of column vectors) for the right nullspace of M, i.e. such that MN
  │  is the zero matrix. If M is an m\times n matrix N will be an n\times \nu
  │  matrix.
```

```julia
p = 23
F = AbstractAlgebra.GF(p)
AAMs = [matrix(F, m) for m in M] # if we implement our own nullspace we don't need this
potential_eigenvalues = Set(F(i) for i in 1:p-1)
eigenvectors = Dict{elem_type(F), Vector}()
Id = identity_matrix(first(AAMs))
for m in view(AAMs, 2:length(AAMs)
    for e in potential_eigenvalues
        nullity, basis = nullspace(m - e*Id)
        if nullity > 0
             # find the "best" eigenvector among basis columns and add it to eigenvectors
             # basis.entries is a standard julia Array
             delete!(potential_eigenvalues, e)
         end
     end
end
```

File: examples/characters.jl [hunk 2]
```diff
@@ -0,0 +1,33 @@
+include("../src/characters.jl")
+
+# Build group ring for the permutation group over 4 elements.
+N = 4
+G = PermGroup(N)
+RG = GroupRing(G)
+
+# Conjugacy classes as all-one elements of GroupRing
+# WTH does all-one elements mean?
```
Comment (kalmarek) [inline examples/characters.jl]: treating elements of RG as function G→R (coefficient ring) it's a characteristic function of a conjugacy class (understood as a simple set). Or a 0-1 coefficients polynomial on a given set of monomials (here: the conjugacy class)

File: examples/characters.jl [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+end
+
+# Multiplication tables for conjugacy classes.
+M = [CCMatrix(ccG, i) for i in 1:length(ccG)];
+
+# First class is the conjugacy class of e. Multiplication with e should not change any conjugacy class.
+@assert M[1] == one(M[1])
+
+# In the sequal we need to do exact arithmetics which is why we work over a finite field. 
+# There are probably smarter ways to find a suitable prime...
```
Comment (kalmarek) [inline examples/characters.jl]: I guess we should use `Primes.jl` for this

**Comment by tweisser** [comment]
Ok, nothing funny with gram_schmidt. It's only that some of the matrices do not have full rank...

**Comment by kalmarek** [comment]
```julia
eigs = let p = 23, Ms = M
    F = AbstractAlgebra.GF(p)
    AAMs = (matrix(F, m) for m in Ms) # if we implement our own nullspace we don't need this
    Id = identity_matrix(first(AAMs))

    potential_eigenvalues = Set(F(i) for i in 1:p-1)
    eigenvectors = Dict{elem_type(F)}()
    

    for m in AAMs
        for e in potential_eigenvalues
            nullity, basis = nullspace(m - e*Id)
            if nullity > 0
                @show e, nullity, basis
                # find the "best" eigenvector among basis columns and add it to eigenvectors
                # basis.entries is a standard julia Array
                entries = basis.entries[:,end]
                eigenvectors[e] = entries
                length(eigenvectors) == size(m,1) && @goto found_eigenbasis
                delete!(potential_eigenvalues, e)
            end
        end
    end
    @label found_eigenbasis
    eigenvectors
end
```

**Comment by kalmarek** [comment]
stage is yours!

**Comment by kalmarek** [comment]
i fucking fixed this! :tada:

**Comment by tweisser** [comment]
WUHU! That's great! Let's meet for a VR-beer to celebrate!

**Comment by kalmarek** [comment]
now?

-------------------------------------------
Marek Kaluba, PhD
Adam Mickiewicz Univeristy
Faculty of Mathematics and Computer Science
ul. Uniwersytetu Poznańskiego 4
61-614 Poznań, Poland
www.amu.edu.pl/~kalmar
-------------------------------------------

**Comment by tweisser** [comment]
waiting for you!

**Pull request closed without merging**